### PR TITLE
[Currency] [RFC] [Critical] Rebased currency party

### LINKED ISSUE #6641: Rounding Issues in Cart
I have an order in my system see image. The unit price * quantity doesn't match the total. When i send the order to paypal i get an error that the totals don't match.

sylius order bug

### PR DESCRIPTION
**Pull request by lchrusciel**
| Q               | A
| --------------- | ---
| Bug fix?        | yes |
| New feature?    | yes |
| BC breaks?      | yes |
| Related tickets | fixes #6641  |
| License         | MIT |

Whole new feature and a lot of improvements to our current logic. As it is the last moment when we could do such a change. With this PR a base currency concept from yaml file. In a replacement, currency will be stored on a channel, which will allow to resolve prices of products, promotions and shipments for each channel separately.  This PR contains changes from  #6814,  #6846, #6820,  #6823, #6844, #6839 and #6831(and maybe a few more).

### CODE REVIEW

File: features/admin/dashboard.feature [hunk 1]
```diff
@@ -12,7 +12,7 @@ Feature: Statistics dashboard in a single channel
         And this product has "Red XL" variant priced at "$40"
         And I am logged in as an administrator
 
-    @ui
+    @ui @todo
```
Comment (pamil) [inline features/admin/dashboard.feature]: 🌮

File: features/admin/dashboard.feature [hunk 2]
```diff
@@ -22,7 +22,7 @@ Feature: Statistics dashboard in a single channel
         And there should be total sales of "$9,025.00"
         And the average order value should be "$1,504.17"
 
-    @ui
+    @ui @todo
```
Comment (pamil) [inline features/admin/dashboard.feature]: 🌮

File: features/admin/dashboard.feature [hunk 3]
```diff
@@ -32,7 +32,7 @@ Feature: Statistics dashboard in a single channel
         And there should be total sales of "$5,241.00"
         And the average order value should be "$1,310.25"
 
-    @ui
+    @ui @todo
```
Comment (pamil) [inline features/admin/dashboard.feature]: 🌮

File: features/cart/shopping_cart/adding_product_to_cart_with_prices_dependent_on_channel.feature [hunk 4]
```diff
@@ -24,22 +21,15 @@ Feature: Adding a product to cart with prices dependent on a channel and a curre
         When I add product "Leprechaun's Gold" to the cart
         Then I should see "Leprechaun's Gold" with unit price "€10.00" in my cart
 
-    @ui
+    @ui @todo
```
Comment (pamil) [inline features/cart/shopping_cart/adding_product_to_cart_with_prices_dependent_on_channel.feature]: 🌮

File: features/cart/shopping_cart/adding_product_to_cart_with_prices_dependent_on_channel.feature [hunk 5]
```diff
@@ -24,22 +21,15 @@ Feature: Adding a product to cart with prices dependent on a channel and a curre
...
     Scenario: Buying a product with proper price after currency change
         Given I change my current channel to "Web-EU"
         When I switch to the "GBP" currency
         And I add product "Leprechaun's Gold" to the cart
-        Then I should see "Leprechaun's Gold" with unit price "£15.00" in my cart
+        Then I should see "Leprechaun's Gold" with unit price "£7.00" in my cart
 
-    @ui
+    @ui @todo
```
Comment (pamil) [inline features/cart/shopping_cart/adding_product_to_cart_with_prices_dependent_on_channel.feature]: 🌮

File: features/product/managing_product_variants/product_variant_validation.feature [hunk 6]
```diff
@@ -15,22 +15,22 @@ Feature: Product variant validation
         When I specify its code as "VODKA_WYBOROWA_PREMIUM"
         But I do not set its price
         And I try to add it
-        Then I should be notified that price is required
+        Then I should be notified that prices in all channels must be defined
         And the "VODKA_WYBOROWA_PREMIUM" variant of the "Wyborowa Vodka" product should not appear in the store
 
-    @ui
+    @ui @todo
```
Comment (pamil) [inline features/product/managing_product_variants/product_variant_validation.feature]: 🌮

File: features/promotion/receiving_discount/receiving_fixed_discount_dependent_on_currency.feature [hunk 7]
```diff
@@ -6,21 +6,21 @@ Feature: Receiving fixed discount dependent on currency on cart
...
         Given the store has currency "USD"
-        And the store has currency "GBP" with exchange rate 0.5
+        And the store has currency "GBP"
         And the store operates on a channel named "Web-US" in "USD" currency
         And that channel allows to shop using "USD" and "GBP" currencies
         And the store has a product "PHP T-Shirt" priced at "$100.00"
         And there is a promotion "Holiday promotion"
 
-    @ui
+    @ui @todo
```
Comment (pamil) [inline features/promotion/receiving_discount/receiving_fixed_discount_dependent_on_currency.feature]: 🌮

File: features/promotion/receiving_discount/receiving_fixed_discount_dependent_on_currency.feature [hunk 8]
```diff
@@ -6,21 +6,21 @@ Feature: Receiving fixed discount dependent on currency on cart
...
     Scenario: Receiving fixed discount in different than base currency for my cart
         Given this promotion gives "$10.00" in base currency or "£8.00" in "GBP" discount to every order
         When I switch to the "GBP" currency
         And I add product "PHP T-Shirt" to the cart
         Then my cart total should be "£42.00"
         And my discount should be "-£8.00"
 
-    @ui
+    @ui @todo
```
Comment (pamil) [inline features/promotion/receiving_discount/receiving_fixed_discount_dependent_on_currency.feature]: 🌮

File: features/account/customer_account/viewing_orders_history/order_is_always_placed_in_a_base_currency_of_a_channel.feature [hunk 9]
```diff
@@ -0,0 +1,30 @@
...
+        And the store ships to "United States"
+        And the store has a zone "United States" with code "US"
+        And this zone has the "United States" country member
+        And the store allows paying with "Cash on Delivery"
+        And the store has "DHL" shipping method with "$20.00" fee within the "US" zone
+        And the store has a product "Angel T-Shirt" priced at "$20.00"
+        And I am a logged in customer
+
+#   This test should pass, but refactoring of context would be needed to fulfill requirements. Because refactoring is postponed this test will be turned on later
+    @ui @todo
```
Comment (pamil) [inline features/account/customer_account/viewing_orders_history/order_is_always_placed_in_a_base_currency_of_a_channel.feature]: 🌮

File: src/Sylius/Behat/Context/Setup/PaymentContext.php [hunk 10]
```diff
@@ -104,6 +104,18 @@ public function storeAllowsPayingForAllChannels($paymentMethodName, array $chann
     }
 
     /**
+     * @Given /^the store allows paying (\w+) for (all channels)$/
+     */
+    public function storeAllowsPayingForAllChannels($paymentMethodName, array $channels)
```
Comment (NeverResponse) [inline src/Sylius/Behat/Context/Setup/PaymentContext.php]: 🚨  duplicated method name

File: src/Sylius/Bundle/CoreBundle/Validator/Constraints/HasAllPricesDefinedValidator.php [hunk 11]
```diff
@@ -0,0 +1,57 @@
...
+    {
+        Assert::isInstanceOf($constraint, HasAllPricesDefined::class);
+
+        $channels = $productVariant->getProduct()->getChannels();
+
+        foreach ($channels as $channel) {
+            /** @var ChannelPricingInterface $channelPricing */
+            $channelPricing = $productVariant->getChannelPricingForChannel($channel);
+            if (null === $channelPricing || null === $channelPricing->getPrice()) {
+                $this->context->addViolationAt('channelPricings', $constraint->getMessage());
```
Comment (pamil) [inline src/Sylius/Bundle/CoreBundle/Validator/Constraints/HasAllPricesDefinedValidator.php]: Deprecated 🗡

File: src/Sylius/Bundle/CoreBundle/Form/Type/Product/ChannelPricingType.php [hunk 12]
```diff
@@ -0,0 +1,51 @@
...
+     */
+    public function buildForm(FormBuilderInterface $builder, array $options)
+    {
+        $builder
+            ->addEventListener(FormEvents::PRE_SET_DATA, function (FormEvent $event) {
+                /** @var ChannelInterface $channel */
+                $channel = $event->getData()->getChannel();
+                $form = $event->getForm();
+
+                $form->add('price', 'sylius_money', [
```
Comment (pamil) [inline src/Sylius/Bundle/CoreBundle/Form/Type/Product/ChannelPricingType.php]: Deprecated 🗡

**Comment by lchrusciel** [comment]
Closing to not block travis

File: features/cart/shopping_cart/changing_dispalay_currency_of_cart.feature [hunk 13]
```diff
@@ -1,12 +1,13 @@
...
+    In order to know estimated price for foreign currencies
     As a Visitor
-    I want to see every cash amount in my chosen currency
+    I want to see every cash amount rounded to my chosen currency
 
     Background:
         Given the store operates on a single channel in "United States"
-        And that channel allows to shop using the "GBP" currency with exchange rate 2.00
+        And that channel allows to shop using the "GBP" currency
+        When the store has an exchange rate 2.0 with source currency "US Dollar" and target currency "British Pound"
```
Comment (pjedrzejewski) [inline features/cart/shopping_cart/changing_dispalay_currency_of_cart.feature]: When the exchange rate of "US Dollar" to "British Pound" is 2.0

Or anything that sounds like human language and not "for-transformers" dialect. :/

File: src/Sylius/Bundle/AdminBundle/Resources/views/Dashboard/_orders.html.twig [hunk 14]
```diff
@@ -20,10 +21,10 @@
                         {{ order.customer.email }}
                     </td>
                     <td>
-                        {{ order.totalQuantity }}
+                        {{ money.format(order.totalQuantity, order.currencyCode) }}
```
Comment (pjedrzejewski) [inline src/Sylius/Bundle/AdminBundle/Resources/views/Dashboard/_orders.html.twig]: Quantity should not be formatted.

File: src/Sylius/Component/Currency/Converter/CurrencyConverter.php [hunk 15]
```diff
@@ -11,71 +11,81 @@
...
     {
-        $this->currencyRepository = $currencyRepository;
+        $this->exchangeRateRepository = $exchangeRateRepository;
     }
 
     /**
      * {@inheritdoc}
      */
-    public function convertFromBase($amount, $targetCurrencyCode)
+    public function convert($amount, $sourceCurrencyCode, $targetCurrencyCode)
```
Comment (pjedrzejewski) [inline src/Sylius/Component/Currency/Converter/CurrencyConverter.php]: If we get pair "USD" & "USD", we should use 1 as exchange rate immediately.

File: src/Sylius/Bundle/AdminBundle/Resources/views/Dashboard/_orders.html.twig (hunk 14 above)
Comment (lchrusciel) [inline src/Sylius/Bundle/AdminBundle/Resources/views/Dashboard/_orders.html.twig]: xD

**Pull request closed without merging**